### PR TITLE
Issue #1792 Splitted up draft functionality from composer

### PR DESCRIPTION
**Pull request by michael-volynets**
This is draft pull request for review only

Issue #1792
Time Spent: 6h so far

### CODE REVIEW

**Comment by michael-volynets** [comment]
@tomholub, what do you think about moving recipients logic to ComposerContacts file?

**Comment by michael-volynets** [comment]
And Could you please again reset ci.compose account? I pushed the changes but hadn't fetched the changes from master and the test failed because of the same thing.

**Comment by michael-volynets** [comment]
Spent about 22 hours in total

**Comment by tomholub** [comment]
> @tomholub, what do you think about moving recipients logic to ComposerContacts file?

Definitely, I expect to expand contact search in near future which will make them even more complicated: #879 so it will be good to have them separate.

**Comment by tomholub** [comment]
> And could you please again reset ci.compose account? I pushed the changes but hadn't fetched the changes from master and the test failed because of the same thing.

Yes, I was thinking this is what was happening. I'll reset it again.

**Comment by tomholub** [comment]
I think I reset it a few days ago, and it did not re-appear. I'll try to re-run this test, maybe it will go through.

**Comment by tomholub** [comment]
Indeed, the test that run after the pull was run after I have already reset it, it's failing for a different reason now:

image

When you open the test result: 
https://flowcrypt.semaphoreci.com/jobs/e24f1b30-2d0f-4509-91a5-0fe7e566dafb
and you search in the page for `FAIL ID`, you will find: `ep9md0zut7`. Search this string in your email and you will see the exact debug file. With that, you can debug it further.

File: extension/chrome/elements/compose.ts [hunk 1]
```diff
@@ -11,12 +11,14 @@ import { Composer } from '../../js/common/composer.js';
...
 import { Google, GoogleAuth } from '../../js/common/api/google.js';
-import { KeyInfo, Contact } from '../../js/common/core/pgp.js';
+import { KeyInfo, Contact, Pgp } from '../../js/common/core/pgp.js';
 import { SendableMsg } from '../../js/common/api/email_provider_api.js';
 import { Assert } from '../../js/common/assert.js';
 import { XssSafeFactory } from '../../js/common/xss_safe_factory.js';
 import { Xss } from '../../js/common/platform/xss.js';
 import { requireOpenpgp } from '../../js/common/platform/require.js';
+import { Keyserver, PubkeySearchResult } from '../../js/common/api/keyserver.js';
+import { PUBKEY_LOOKUP_RESULT_FAIL } from '../../js/common/composer/interfaces/comopserr-errors.js';
```
Comment (tomholub) [inline extension/chrome/elements/compose.ts]: `comopserr-errors` typo

File: extension/chrome/elements/reply_pubkey_mismatch.ts [hunk 2]
```diff
@@ -30,7 +31,43 @@ Catch.try(async () => {
...
+    emailEroviderSearchContacts: (query, knownContacts, multiCb) => multiCb({ new: [], all: [] }),
+    emailProviderDetermineReplyMsgHeaderVariables: () => Promise.resolve(undefined),
+    emailProviderExtractArmoredBlock: () => Promise.resolve(''),
+    renderReinsertReplyBox: () => Promise.resolve(),
+    renderFooterDialog: () => undefined,
+    renderAddPubkeyDialog: () => undefined,
+    renderHelpDialog: () => undefined,
+    renderSendingAddrDialog: () => undefined,
+    closeMsg: () => undefined,
+    factoryAtt: (att) => `<div>${att.name}</div>`,
```
Comment (tomholub) [inline extension/chrome/elements/reply_pubkey_mismatch.ts]: Just newly noticed this. `att.name` should be Xss escaped.

File: extension/js/common/composer/composer-quote.ts [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+
+export class ComposerQuote extends ComposerComponent {
+    private messageToReplyOrForward: MessageToReplyOrForward | undefined;
+    private msgExpandingHTMLPart: string | undefined;
+
+    get expandingHTMLPart(): string | undefined {
+        return this.msgExpandingHTMLPart;
+    }
+
+    constructor(composer: Composer, urlParams: ComposerUrlParams) {
```
Comment (tomholub) [inline extension/js/common/composer/composer-quote.ts]: Maybe this constructor is unnecessary? If if you want to keep it in for code clarity, you can.

File: extension/js/common/composer/composer-draft.ts [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+import { Buf } from '../core/buf.js';
+import { Pgp, PgpMsg } from '../core/pgp.js';
+import { Api, AjaxError } from '../api/api.js';
+import { BrowserMsg } from '../extension.js';
+import { Catch } from '../platform/catch.js';
+import { Store } from '../platform/store.js';
+import { Str } from '../core/common.js';
+import { Composer } from '../composer.js';
+import { ComposerUrlParams } from './interfaces/composer-types.js';
+import { ComposerComponent } from './interfaces/comopser-component.js';
```
Comment (tomholub) [inline extension/js/common/composer/composer-draft.ts]: typo `comopser-component`

Comment (michael-volynets) [reply extension/js/common/composer/composer-draft.ts]: so many typos, sorry

File: extension/js/common/composer/composer-quote.ts (hunk 3 above)
Comment (michael-volynets) [reply extension/js/common/composer/composer-quote.ts]: Agree, this one is unnecessary

File: extension/js/common/composer/composer-draft.ts (hunk 4 above)
Comment (tomholub) [reply extension/js/common/composer/composer-draft.ts]: Maybe can install some word-checking plugin. IntelliJ will always underline typos everywhere. There must be something similar for vscode. Then when we add a word to the dictionary of the plugin, the dict file should be check into the repo so everyone can benefit.

File: extension/chrome/elements/reply_pubkey_mismatch.ts [hunk 5]
```diff
@@ -63,7 +63,7 @@ Catch.try(async () => {
     renderHelpDialog: () => undefined,
     renderSendingAddrDialog: () => undefined,
     closeMsg: () => undefined,
-    factoryAtt: (att) => `<div>${att.name}</div>`,
+    factoryAtt: (att) => `<div>${att.name}</div>`, // xss-escaped
```
Comment (tomholub) [inline extension/chrome/elements/reply_pubkey_mismatch.ts]: I meant actually escape it: `<div>${Xss.escape(att.name)}</div>`

Comment (michael-volynets) [reply extension/chrome/elements/reply_pubkey_mismatch.ts]: oh, sorry, corrected

**Pull request merged**